File: content/problems/3067-count-pairs-of-connectable-servers-in-a-weighted-tree-network/solutions/enumerate-centers-and-count-branches/solution.cpp
class Solution {
public:
    vector<int> countPairsOfConnectableServers(vector<vector<int>>& edges,int signalSpeed){
        int n=edges.size()+1;
        vector<vector<pair<int,int>>> graph(n);
        for(auto& edge:edges){graph[edge[0]].push_back({edge[1],edge[2]});graph[edge[1]].push_back({edge[0],edge[2]});}
        vector<int> answer(n);
        function<int(int,int,long long)> count=[&](int node,int parent,long long distance){
            int result=distance%signalSpeed==0;
            for(auto [next,weight]:graph[node])if(next!=parent)result+=count(next,node,distance+weight);
            return result;
        };
        for(int server=0;server<n;++server){
            int previous=0;
            for(auto [next,weight]:graph[server]){
                int current=count(next,server,weight);
                answer[server]+=previous*current;
                previous+=current;
            }
        }
        return answer;
    }
};

File: content/problems/3067-count-pairs-of-connectable-servers-in-a-weighted-tree-network/solutions/enumerate-centers-and-count-branches/Solution.java
class Solution {
    private int count(int node,int parent,long distance,int speed,List<int[]>[] graph){
        int result=distance% speed==0?1:0;
        for(int[] edge:graph[node])if(edge[0]!=parent)result+=count(edge[0],node,distance+edge[1],speed,graph);
        return result;
    }
    public int[] countPairsOfConnectableServers(int[][] edges,int signalSpeed){
        int n=edges.length+1;
        List<int[]>[] graph=new ArrayList[n];
        for(int node=0;node<n;node++)graph[node]=new ArrayList<>();
        for(int[] edge:edges){graph[edge[0]].add(new int[]{edge[1],edge[2]});graph[edge[1]].add(new int[]{edge[0],edge[2]});}
        int[] answer=new int[n];
        for(int server=0;server<n;server++){
            int previous=0;
            for(int[] edge:graph[server]){
                int current=count(edge[0],server,edge[1],signalSpeed,graph);
                answer[server]+=previous*current;
                previous+=current;
            }
        }
        return answer;
    }
}
